viewing_party/party.py

```diff
@@ -0,0 +1,243 @@
+                    friends_watched_set.add(indiv_dict["title"])
+
+    only_user_watched_set = user_watched_set.difference(friends_watched_set)
+
+    for item in only_user_watched_set:
+        only_user_watched.append({"title" : item})
+
+    return only_user_watched
+                    
+    
+
+#wave 3, part 2
+def get_friends_unique_watched(user_data):
+    """Takes dict, returns list of dicts that only friends have seen."""
+
+    user_watched_set = set()
+    friends_watched_set = set()
+    only_friends_watched_set = set()
+    only_friends_watched = []
+
+    for movie_user_watched in user_data["watched"]:
+        user_watched_set.add(movie_user_watched["title"])
+        
+    for friends_watched_lists in user_data["friends"]:
+        for movies_friends_watched_list in friends_watched_lists.values():
+            for indiv_dict in movies_friends_watched_list:
+                friends_watched_set.add(indiv_dict["title"])
+
+    only_friends_watched_set = friends_watched_set.difference(user_watched_set)
+
+    for item in only_friends_watched_set:
+        only_friends_watched.append({"title" : item})
+
+    return only_friends_watched
+
+
+
+
+################ WAVE 4 ##############
+
+def get_available_recs(user_data):
+    """Takes dict, returns list of movie recommendations."""
+
+    friends_watched_list = get_friends_unique_watched(user_data)
+    friends_watched_list_titles = []
+    rec_list = []
+    
+    for item in friends_watched_list:
+        friends_watched_list_titles.append(item["title"])
+
+    for watched_lists in user_data["friends"]:
+        for watched_dict in watched_lists["watched"]:
+            if (watched_dict["host"] in user_data["subscriptions"]) \
+                and (watched_dict not in rec_list) \
+                and (watched_dict["title"] in friends_watched_list_titles):
+                    rec_list.append(watched_dict)
+    
+    return rec_list
+
+
+
+
+
+################ WAVE 5 ##############
+
+#wave 5, part 1
+def get_new_rec_by_genre(user_data):
+    """Takes dict, returns list of recs that match most pop genre."""
+
+    recs_by_genre_list = []
+    fav_genre = get_most_watched_genre(user_data)
+   
+    for items in user_data["friends"]:
+        for movie_dict in items["watched"]:
+            if ((movie_dict["title"] not in user_data["watched"]) \
+            and (movie_dict["genre"] == fav_genre) \
+            and (movie_dict not in recs_by_genre_list)):
+                recs_by_genre_list.append(movie_dict)
+
+    return recs_by_genre_list
+
+
+#wave 5, part 2
+def get_rec_from_favorites(user_data):
+    """Takes dict, returns list of recommendations."""
+  
+    movie_recs = []
+    friends_watched_set = set()
+  
+    for friends_watched_lists in user_data["friends"]:
+        for movies_friends_watched_list in friends_watched_lists.values():
+            for indiv_dict in movies_friends_watched_list:
+                friends_watched_set.add(indiv_dict["title"])
```

Comment:
Here again, we could simplify the looping as in `get_unique_watched`. No need for the `.values()` call.